hbase/src/main/scala/akka/stream/alpakka/hbase/internal/HBaseFlowStage.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.hbase.internal
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.stream.alpakka.hbase.HTableSettings
+import akka.stream.stage._
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+import org.apache.hadoop.hbase.client.Table
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success, Try }
+
+private[hbase] class HBaseFlowStage[A](settings: HTableSettings[A])
+    extends GraphStage[FlowShape[A, Future[A]]]
+    with HBaseCapabilities {
+
+  private val in = Inlet[A]("messages")
+  private val out = Outlet[Future[A]]("result")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
+
+    implicit val connection = connect(settings.conf)
+
+    val table: Table = getOrCreateTable(settings.tableName, settings.columnFamilies)
+
+    new GraphStageLogic(shape) with StageLogging {
+      val awaitingConfirmation = new AtomicInteger(0)
+      @volatile var inIsClosed = false
+      var completionState: Option[Try[Unit]] = None
+
+      setHandler(out, new OutHandler {
+        override def onPull() =
+          tryPull(in)
+      })
+
+      setHandler(in, new InHandler {
+        override def onPush() = {
+          val msg = grab(in)
+          val r = Promise[A]
+
+          awaitingConfirmation.incrementAndGet()
+
+          Future {
+            table.put(settings.converter.put(msg))
+          }.map { r =>
+            if (awaitingConfirmation.decrementAndGet() == 0 && inIsClosed)
+              checkForCompletion()
+            r
+          } onComplete {
+            case Success(e) =>
+              r.success(msg)
+            case Failure(e) =>
+              r.failure(e)
+          }
+
+          push(out, r.future)
+        }
+
+        override def onUpstreamFinish() = {
+          inIsClosed = true
+          completionState = Some(Success(()))
+          checkForCompletion()
+        }
+
+        override def onUpstreamFailure(ex: Throwable) = {
+          inIsClosed = true
+          completionState = Some(Failure(ex))
+          checkForCompletion()
+        }
+
+      })
+
+      def checkForCompletion() =
+        if (isClosed(in) && awaitingConfirmation.get == 0) {
+          completionState match {
+            case Some(Success(_)) => completeStage()
+            case Some(Failure(ex)) => failStage(ex)
+            case None => failStage(new IllegalStateException("Stage completed, but there is no info about status"))
+          }
+        }
+
+      val checkForCompletionCB = getAsyncCallback[Unit] { _ =>
+        checkForCompletion()
+      }
+
+      override def postStop() = {
+        log.debug("Stage completed")
+        try {
+          table.close()
+          connection.close()
```

Comment:
shouldn't these be separate tries, if table.close throws this will leave the connection unclosed?